pydantic/funcs.py

```diff
@@ -0,0 +1,226 @@
+from __future__ import annotations
+
+import sys
+from typing import TYPE_CHECKING, Any, Dict, Generic, Iterable, Set, TypeVar, Union, overload
+
+from pydantic_core import CoreConfig, SchemaSerializer, SchemaValidator, core_schema
+from typing_extensions import Literal
+
+from pydantic.json_schema import DEFAULT_REF_TEMPLATE, GenerateJsonSchema
+
+from ._internal import (
+    _generate_schema,
+    _typing_extra,
+)
+
+T = TypeVar('T')
+
+if TYPE_CHECKING:
+    # should be `set[int] | set[str] | dict[int, IncEx] | dict[str, IncEx] | None`, but mypy can't cope
+    IncEx = Union[Set[int], Set[str], Dict[int, Any], Dict[str, Any]]
+
+
+class AnalyzedType(Generic[T]):
+    def __init__(self, core_schema: core_schema.CoreSchema, core_config: CoreConfig) -> None:
+        self.core_schema = core_schema
+        self.core_config = core_config
+        self._validator: SchemaValidator | None = None
+        self._schema: SchemaSerializer | None = None
+
+    @property
```

Comment:
any reason not to use `functools.cached_property` here? I think it cleans things up a bit as you can drop the `_valdiator` and `_schema` properties and don't need the if checks in the property. I guess in principle it's faster too as it sets an instance attribute when called.

Reply:
Isn’t it only available in 3.8?